wheel/python/clvm_rs/program.py

```diff
@@ -0,0 +1,345 @@
+                prog_bytes, args_bytes, max_cost, flags
+            )
+            r = self.wrap(lazy_node)
+        except ValueError as ve:
+            raise EvalError(ve.args[0], self.wrap(ve.args[1]))
+        return cost, r
+
+    def run(self, args) -> "Program":
+        cost, r = self.run_with_cost(args, MAX_COST)
+        return r
+
+    def curry(self, *args) -> "Program":
+        """
+        Replicates the curry function from clvm_tools, taking advantage of *args
+        being a list.  We iterate through args in reverse building the code to
+        create a clvm list.
+
+        Given arguments to a function addressable by the '1' reference in clvm
+
+        fixed_args = 1
+
+        Each arg is prepended as fixed_args = (c (q . arg) fixed_args)
+
+        The resulting argument list is interpreted with apply (2)
+
+        (a (q . self) rest)
+
+        Resulting in a function which places its own arguments after those
+        curried in in the form of a proper list.
+        """
+        return self.to(self.curry_treehasher.curry(self, *args))
+
+    def uncurry(self) -> Tuple[Program, Optional[List[Program]]]:
+        """
+        uncurry the given program
+
+        returns `mod, [arg1, arg2, ...]`
+
+        if the program is not a valid curry, return `self, NULL`
+
+        This distinguishes it from the case of a valid curry of 0 arguments
+        (which is rather pointless but possible), which returns `self, []`
+        """
+        mod, args = self.curry_treehasher.uncurry(self)
+        p_args = args if args is None else [self.to(_) for _ in args]
+        return self.to(mod), p_args
+
+    def curry_hash(self, *args: bytes32) -> bytes32:
+        """
+        Return a puzzle hash that would be created if you curried this puzzle
+        with arguments that have the given hashes.
+
+        In other words,
+
+        ```
+        c1 = self.curry(arg1, arg2, arg3).tree_hash()
+        c2 = self.curry_hash(arg1.tree_hash(), arg2.tree_hash(), arg3.tree_hash())
+        assert c1 == c2  # they will be the same
+        ```
+
+        This looks useless to the unitiated, but sometimes you'll need a puzzle
+        hash where you don't actually know the contents of a clvm subtree -- just its
+        hash. This lets you calculate the puzzle hash with hidden information.
+        """
+        curry_treehasher = self.curry_treehasher
+        quoted_mod_hash = curry_treehasher.calculate_hash_of_quoted_mod_hash(
+            self.tree_hash()
+        )
+        return curry_treehasher.curry_and_treehash(quoted_mod_hash, *args)
+
+    def as_int(self) -> int:
+        v = self.as_atom()
+        if v is None:
+            raise ValueError("can't cast pair to int")
+        return int_from_bytes(v)
+
+    def as_iter(self) -> Iterator[Program]:
+        v = self
+        while v.pair:
+            yield v.pair[0]
+            v = v.pair[1]
+
+    def as_atom_iter(self) -> Iterator[bytes]:
```

Comment:
This API has been around for a long time. It was sort of an evolution from `.to_python` (which can be problematically slow for deep trees) to `.as_atom_list` and `.as_atom_iter`. Once `.at` was discovered, it seems clear it's the best API and these others are not useful.

I've gradually grown to dislike exceptions in normal flow because they're can't be expressed in type hints. I think my experience with rust and it's uber-awesome type system has influenced me (it has no exceptions, just good patterns for error handling). I mostly envision this code to be used as a convenience in hack test code.

I should probably make `.as_int` return `Optional[int]` rather than raise (`__int__` will have to raise though).

Maybe I should just remove them.

Reply:
Looking in `chia-blockchain`, one of these is used in one place (`parse_sexp_to_condition` which is obsolete anyway) and the other is not used at all, and I think they are both bad API compared to `.at`, so I'm going to remove them.

Reminds me: I want a second version of `.at` (maybe `.at_many`) which takes multiple args for even more convenience. So instead of

`if p.at("f") == 2 and p.at("rf") == 5 and p.at("rrfrf") == 10:`

you write

`if p.at_many("f", "rf", "rrfrf") == (2, 5, 10):`

Reply:
I removed the `.as_iter` and `.as_list` methods and added `.at_many`.

Reply:
Do you have plans to implement `.to_python()` or is there a way I can emulate that behavior with the current API?

Reply:
No, as `.to_python` on untrusted clvm is a DOS vector. The primary use case seems to be conditions, which are, generally speaking, untrusted, since if they were created locally, you'd know what they were anyway.

Reply:
The main use case for me is simply to do `puzzle_hash: bytes32 = bytes32(some_prog.to_python())`.  How would I do that in this class?

Reply:
`some_prog.atom`

Reply:
Way shorter too.

Reply:
I could be mistaken but I thought `.atom` would return to me the same as `bytes(some_prog)` which includes the `a0` length prefix.  So now my code would have to be `puzzle_hash: bytes32 = bytes32(some_prog.atom[1:])` which is a bit of magic knowledge that feels bad.  It's a very small difference granted "I want the bytes" vs "I want the bytes w/o the length prefix", but it does seem like having a method/property for that would be good.  Even just `.to_python` that raises when the program isn't an atom would be good.